src/NuGet.Clients/NuGet.SolutionRestoreManager/InfoBarService.cs

```diff
@@ -0,0 +1,165 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.ComponentModel.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio;
+using Microsoft.VisualStudio.Imaging;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using NuGet.VisualStudio;
+using NuGet.VisualStudio.Telemetry;
+
+#nullable enable
+
+namespace NuGet.SolutionRestoreManager
+{
+    [Export(typeof(IInfoBarService))]
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    public class InfoBarService : IInfoBarService, IVsInfoBarUIEvents
+    {
+        IAsyncServiceProvider _asyncServiceProvider = AsyncServiceProvider.GlobalProvider;
+        private IVsInfoBarUIElement? _infoBarUIElement;
+        private bool _visible;
+        private bool _closed = false;
+        private bool _closeFromHide = false;
+        private uint? _eventCookie;
+
+        [Import]
+        private Lazy<IPMUIStarter>? PackageManagerUIStarter { get; set; }
+
+        public async Task ShowAsync(CancellationToken cancellationToken)
+        {
+            await CreateAndShowInfoBarAsync(cancellationToken);
+        }
+
+        public async Task RemoveAsync(CancellationToken cancellationToken)
+        {
+            await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
+            if (!_visible)
+            {
+                return;
+            }
+
+            _closeFromHide = true;
+
+            try
+            {
+                if (_eventCookie.HasValue)
+                {
+                    _infoBarUIElement?.Close();
+                    _infoBarUIElement?.Unadvise(_eventCookie.Value);
+                    _infoBarUIElement = null;
+                }
+            }
+            finally
+            {
+                _closeFromHide = false;
+            }
+        }
+
+        protected InfoBarModel GetInfoBarModel()
+        {
+            return new InfoBarModel(
```

Comment:
[InfoBarModel(String, IEnumerable<IVsInfoBarActionItem>, ImageMoniker, Boolean)](https://learn.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.shell.infobarmodel.-ctor?view=visualstudiosdk-2022#microsoft-visualstudio-shell-infobarmodel-ctor(system-string-system-collections-generic-ienumerable((microsoft-visualstudio-shell-interop-ivsinfobaractionitem))-microsoft-visualstudio-imaging-interop-imagemoniker-system-boolean)) might be appropriate here because there is no need to create a list for `IVsInfoBarTextSpan`.